build: Add support for Android NDK r22+

hebasto opened the pull request:
This is required to support [`std::filesystem`](https://github.com/android/ndk/wiki/Changelog-r22#changes) on Android (see #20744).

Fixes #22074.

hebasto committed:
build: Add support for Android NDK r22+

hebasto committed:
ci: Bump Android NDK to r22 which supports std::filesystem

fanquake commented:
Concept ACK - Thanks for looking into this, and the patching looks reasonable. As I [said in #22074](https://github.com/bitcoin/bitcoin/issues/22074#issuecomment-871337482):
> I can't imagine we are going to block migrating to std::filesystem based on our Qt version being too old, to support building against a new enough version of the Android NDK to support std::filesystem. I'd rather just drop the Android build until we've moved on to a newer version of Qt.

So if this is a solution for that 👍

hebasto commented:
cc @icota

icota commented on depends/patches/qt/support_new_android_ndks.patch:
```diff
@@ -0,0 +1,123 @@
++    QMAKE_CFLAGS = -target x86_64-linux-android$$replace(ANDROID_PLATFORM, "android-", "")
+ else: equals(ANDROID_TARGET_ARCH, mips): \
+-    QMAKE_CFLAGS += -target mipsel-none-linux-android
++    QMAKE_CFLAGS += -target mipsel-linux-android$$replace(ANDROID_PLATFORM, "android-", "")
```
`+=` operator here unlike `=` in other `TARGET_ARCHES` above? 

Personally I'd get rid of `mips` as it's no longer suppored and to make the patchfile that much smaller.

hebasto replied:
mips acrhs are removed in Qt 5.13 (https://github.com/qt/qtbase/commit/854156dd07ca5d50db9be091d34f68b9ca02ff21), but that commit is orthogonal to the issues solved here.

hebasto replied:
> `+=` operator here unlike `=` in other `TARGET_ARCHES` above?

Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/22469#issuecomment-881861647).

fanquake replied:
> Personally I'd get rid of mips as it's no longer suppored and to make the patchfile that much smaller.

Inclined to agree with @icota here. In any case, if mips support was removed in 5.13, and the patch here is from 5.14, why does it contain anything mips related? If you've added that to the patch yourself, why isn't that mentioned / documented anywhere?

hebasto replied:
Agree.

> Personally I'd get rid of `mips` as it's no longer suppored and to make the patchfile that much smaller.

> Inclined to agree with @icota here.

[Updated](https://github.com/bitcoin/bitcoin/pull/22469#issuecomment-882055056).

hebasto commented:
Updated 4d7fa17e34947293d5ed542128ee60cef4e99b17 -> 464c46d4dea028c8e9bcfc2e0abfd171238075ea ([pr22469.01](https://github.com/hebasto/bitcoin/commits/pr22469.01) -> [pr22469.02](https://github.com/hebasto/bitcoin/commits/pr22469.02), [diff](https://github.com/hebasto/bitcoin/compare/pr22469.01..pr22469.02)).

Addressed @icota's [comment](https://github.com/bitcoin/bitcoin/pull/22469#discussion_r671644196):
> `+=` operator here unlike `=` in other `TARGET_ARCHES` above?

icota commented:
tACK https://github.com/bitcoin/bitcoin/commit/464c46d4dea028c8e9bcfc2e0abfd171238075ea
Built on CI, APK works fine on a device. Thanks @hebasto!

hebasto commented:
Updated 464c46d4dea028c8e9bcfc2e0abfd171238075ea -> acaac6e86a9e808244d9c69a59ab3c2d8e34cad6 ([pr22469.02](https://github.com/hebasto/bitcoin/commits/pr22469.02) -> [pr22469.03](https://github.com/hebasto/bitcoin/commits/pr22469.03)):

- rebased due to the conflict with #20641
- addressed @icota's and @fanquake's [comments](https://github.com/bitcoin/bitcoin/pull/22469#discussion_r671786517)

BlockMechanic commented:
Nice ! This is a much needed change, will test on 23 to see if it works as well

icota commented:
re-tACK https://github.com/bitcoin/bitcoin/commit/acaac6e86a9e808244d9c69a59ab3c2d8e34cad6

fanquake commented:
Guix builds:
```bash
d38cd187f6ed1e7a9864c4f140589a1e7ea4f48d7928c96de98583f9b6f1b2a7  guix-build-acaac6e86a9e/output/aarch64-linux-gnu/SHA256SUMS.part
921c3615fc2e5acbc5a035bf6f2e24a0f70a5714ec1d80515c77436e1b5ff61d  guix-build-acaac6e86a9e/output/aarch64-linux-gnu/bitcoin-acaac6e86a9e-aarch64-linux-gnu-debug.tar.gz
19cf6afd7f3e37d63584d38b3abac6a3a958264f310793fc10076af184577874  guix-build-acaac6e86a9e/output/aarch64-linux-gnu/bitcoin-acaac6e86a9e-aarch64-linux-gnu.tar.gz
8dd20df0726a79c61ce1f4f57479dc0ed62919409c738d1ccdd36e411bf78102  guix-build-acaac6e86a9e/output/arm-linux-gnueabihf/SHA256SUMS.part
5d2c56cf68f035a46ecece92e7d9bbfa5e35b68c6589fba149542e474adb7509  guix-build-acaac6e86a9e/output/arm-linux-gnueabihf/bitcoin-acaac6e86a9e-arm-linux-gnueabihf-debug.tar.gz
90e381a9ca1b3bed35f09d91fdc99ce0bfbafd3ba888c514ee1d82be92884f37  guix-build-acaac6e86a9e/output/arm-linux-gnueabihf/bitcoin-acaac6e86a9e-arm-linux-gnueabihf.tar.gz
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  guix-build-acaac6e86a9e/output/dist-archive/SKIPATTEST.TAG
a70163b13242361b0c74e8341e2a7148bde92407f6c45edf784c4598b05daa09  guix-build-acaac6e86a9e/output/dist-archive/bitcoin-acaac6e86a9e.tar.gz
f3c256dff1d89afb1dbf6b9ea61d51eb867a07ec452da30c88508685c4736b75  guix-build-acaac6e86a9e/output/powerpc64-linux-gnu/SHA256SUMS.part
e1fc374fd751ce7339ef1e2d4adcd55704aa2d84fab4e954eea9b6dcf258663c  guix-build-acaac6e86a9e/output/powerpc64-linux-gnu/bitcoin-acaac6e86a9e-powerpc64-linux-gnu-debug.tar.gz
f3b87be252488a6464d18e01d5cb75e83b1f11f9413f744da6297c0c474a333e  guix-build-acaac6e86a9e/output/powerpc64-linux-gnu/bitcoin-acaac6e86a9e-powerpc64-linux-gnu.tar.gz
191d88ca173730cabd2a2218e87306697e92f0ccb5b279d1d2cc5f1f7ac16c97  guix-build-acaac6e86a9e/output/powerpc64le-linux-gnu/SHA256SUMS.part
bf52b6dc51cdbbcaae6a9d9f6e397612520e8ba211cd779856f5e85e6d7e65ff  guix-build-acaac6e86a9e/output/powerpc64le-linux-gnu/bitcoin-acaac6e86a9e-powerpc64le-linux-gnu-debug.tar.gz
9f4740eb595ebf2263a526915e28eaf2c9ee16f5ce4b1cc8894fa6ac22f35647  guix-build-acaac6e86a9e/output/powerpc64le-linux-gnu/bitcoin-acaac6e86a9e-powerpc64le-linux-gnu.tar.gz
0c5397923650ac212b314370318ed67843f65f4b8e2b16f50c23a777f4d33d62  guix-build-acaac6e86a9e/output/riscv64-linux-gnu/SHA256SUMS.part
7f975c1c9b4d2e7944a85c8cacc34c328859812ba6e68bfdb05f507fc4962603  guix-build-acaac6e86a9e/output/riscv64-linux-gnu/bitcoin-acaac6e86a9e-riscv64-linux-gnu-debug.tar.gz
ea2751fc2f60539dfa3329ba859401d97a040e2c1f42eb93cfa64757d110a3c8  guix-build-acaac6e86a9e/output/riscv64-linux-gnu/bitcoin-acaac6e86a9e-riscv64-linux-gnu.tar.gz
a8001e596f1050a89ed0965fba738b4d72c48262dfa4f0649eb85c75d6ea992a  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/SHA256SUMS.part
55f8b10c4f909402b5ae0f7d976a521ca4311f971262e81fc2a862c2c2ef6411  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/bitcoin-acaac6e86a9e-osx-unsigned.dmg
a30aa30be3f34bfde9e7fec556c910cfc66b124608c9ed35cbc139a9e99be35e  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/bitcoin-acaac6e86a9e-osx-unsigned.tar.gz
4fdee4745af8a2a7d6f52e8ad9fe1b6809f0c5a02f0dcfd0432eb807a205e536  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/bitcoin-acaac6e86a9e-osx64.tar.gz
415994004fb9b8f70e73927324e72e899050cdb74ad5d377daa725844aa5287c  guix-build-acaac6e86a9e/output/x86_64-linux-gnu/SHA256SUMS.part
67aab9cdd3608135c33ab3f9801ee9b1e554229cfdd692c0664c6f06f52e55ba  guix-build-acaac6e86a9e/output/x86_64-linux-gnu/bitcoin-acaac6e86a9e-x86_64-linux-gnu-debug.tar.gz
c9f7f0465cfd12ae2a74b7734008b6a0a56b5be4392e694bc958a0c2671153bf  guix-build-acaac6e86a9e/output/x86_64-linux-gnu/bitcoin-acaac6e86a9e-x86_64-linux-gnu.tar.gz
34acf28dbdd450fc3ae969e38874cc99557dd72a1ee54291d88e81a95b1fe58a  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/SHA256SUMS.part
f8d7f4ecbee65750ff60b8093acf868c5246251fd6897f7e8a4b310285b535f4  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win-unsigned.tar.gz
c411cf9d1ef1a9e02df4fc2306cfb1dd40324205f84f5c860cbf849703ab9d23  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win64-debug.zip
cdb7d1422a6986329f06e1fd64ecfce751e5a892f4c5b316fc15ff4f2655623c  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win64-setup-unsigned.exe
9dd9696f85f91f5970e76d0d65bb929769c601dfdc1505a2b89a92909c1bc724  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win64.zip
```

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
d38cd187f6ed1e7a9864c4f140589a1e7ea4f48d7928c96de98583f9b6f1b2a7  guix-build-acaac6e86a9e/output/aarch64-linux-gnu/SHA256SUMS.part
921c3615fc2e5acbc5a035bf6f2e24a0f70a5714ec1d80515c77436e1b5ff61d  guix-build-acaac6e86a9e/output/aarch64-linux-gnu/bitcoin-acaac6e86a9e-aarch64-linux-gnu-debug.tar.gz
19cf6afd7f3e37d63584d38b3abac6a3a958264f310793fc10076af184577874  guix-build-acaac6e86a9e/output/aarch64-linux-gnu/bitcoin-acaac6e86a9e-aarch64-linux-gnu.tar.gz
8dd20df0726a79c61ce1f4f57479dc0ed62919409c738d1ccdd36e411bf78102  guix-build-acaac6e86a9e/output/arm-linux-gnueabihf/SHA256SUMS.part
5d2c56cf68f035a46ecece92e7d9bbfa5e35b68c6589fba149542e474adb7509  guix-build-acaac6e86a9e/output/arm-linux-gnueabihf/bitcoin-acaac6e86a9e-arm-linux-gnueabihf-debug.tar.gz
90e381a9ca1b3bed35f09d91fdc99ce0bfbafd3ba888c514ee1d82be92884f37  guix-build-acaac6e86a9e/output/arm-linux-gnueabihf/bitcoin-acaac6e86a9e-arm-linux-gnueabihf.tar.gz
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  guix-build-acaac6e86a9e/output/dist-archive/SKIPATTEST.TAG
a70163b13242361b0c74e8341e2a7148bde92407f6c45edf784c4598b05daa09  guix-build-acaac6e86a9e/output/dist-archive/bitcoin-acaac6e86a9e.tar.gz
f3c256dff1d89afb1dbf6b9ea61d51eb867a07ec452da30c88508685c4736b75  guix-build-acaac6e86a9e/output/powerpc64-linux-gnu/SHA256SUMS.part
e1fc374fd751ce7339ef1e2d4adcd55704aa2d84fab4e954eea9b6dcf258663c  guix-build-acaac6e86a9e/output/powerpc64-linux-gnu/bitcoin-acaac6e86a9e-powerpc64-linux-gnu-debug.tar.gz
f3b87be252488a6464d18e01d5cb75e83b1f11f9413f744da6297c0c474a333e  guix-build-acaac6e86a9e/output/powerpc64-linux-gnu/bitcoin-acaac6e86a9e-powerpc64-linux-gnu.tar.gz
191d88ca173730cabd2a2218e87306697e92f0ccb5b279d1d2cc5f1f7ac16c97  guix-build-acaac6e86a9e/output/powerpc64le-linux-gnu/SHA256SUMS.part
bf52b6dc51cdbbcaae6a9d9f6e397612520e8ba211cd779856f5e85e6d7e65ff  guix-build-acaac6e86a9e/output/powerpc64le-linux-gnu/bitcoin-acaac6e86a9e-powerpc64le-linux-gnu-debug.tar.gz
9f4740eb595ebf2263a526915e28eaf2c9ee16f5ce4b1cc8894fa6ac22f35647  guix-build-acaac6e86a9e/output/powerpc64le-linux-gnu/bitcoin-acaac6e86a9e-powerpc64le-linux-gnu.tar.gz
0c5397923650ac212b314370318ed67843f65f4b8e2b16f50c23a777f4d33d62  guix-build-acaac6e86a9e/output/riscv64-linux-gnu/SHA256SUMS.part
7f975c1c9b4d2e7944a85c8cacc34c328859812ba6e68bfdb05f507fc4962603  guix-build-acaac6e86a9e/output/riscv64-linux-gnu/bitcoin-acaac6e86a9e-riscv64-linux-gnu-debug.tar.gz
ea2751fc2f60539dfa3329ba859401d97a040e2c1f42eb93cfa64757d110a3c8  guix-build-acaac6e86a9e/output/riscv64-linux-gnu/bitcoin-acaac6e86a9e-riscv64-linux-gnu.tar.gz
a8001e596f1050a89ed0965fba738b4d72c48262dfa4f0649eb85c75d6ea992a  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/SHA256SUMS.part
55f8b10c4f909402b5ae0f7d976a521ca4311f971262e81fc2a862c2c2ef6411  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/bitcoin-acaac6e86a9e-osx-unsigned.dmg
a30aa30be3f34bfde9e7fec556c910cfc66b124608c9ed35cbc139a9e99be35e  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/bitcoin-acaac6e86a9e-osx-unsigned.tar.gz
4fdee4745af8a2a7d6f52e8ad9fe1b6809f0c5a02f0dcfd0432eb807a205e536  guix-build-acaac6e86a9e/output/x86_64-apple-darwin18/bitcoin-acaac6e86a9e-osx64.tar.gz
415994004fb9b8f70e73927324e72e899050cdb74ad5d377daa725844aa5287c  guix-build-acaac6e86a9e/output/x86_64-linux-gnu/SHA256SUMS.part
67aab9cdd3608135c33ab3f9801ee9b1e554229cfdd692c0664c6f06f52e55ba  guix-build-acaac6e86a9e/output/x86_64-linux-gnu/bitcoin-acaac6e86a9e-x86_64-linux-gnu-debug.tar.gz
c9f7f0465cfd12ae2a74b7734008b6a0a56b5be4392e694bc958a0c2671153bf  guix-build-acaac6e86a9e/output/x86_64-linux-gnu/bitcoin-acaac6e86a9e-x86_64-linux-gnu.tar.gz
34acf28dbdd450fc3ae969e38874cc99557dd72a1ee54291d88e81a95b1fe58a  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/SHA256SUMS.part
f8d7f4ecbee65750ff60b8093acf868c5246251fd6897f7e8a4b310285b535f4  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win-unsigned.tar.gz
c411cf9d1ef1a9e02df4fc2306cfb1dd40324205f84f5c860cbf849703ab9d23  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win64-debug.zip
cdb7d1422a6986329f06e1fd64ecfce751e5a892f4c5b316fc15ff4f2655623c  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win64-setup-unsigned.exe
9dd9696f85f91f5970e76d0d65bb929769c601dfdc1505a2b89a92909c1bc724  guix-build-acaac6e86a9e/output/x86_64-w64-mingw32/bitcoin-acaac6e86a9e-win64.zip
```
